### PR TITLE
Refactor region

### PR DESCRIPTION
**Pull request by samjrholt**
@lang-m I've made the initial critical changes we discussed and added tests.

There are two main issues that I have at the moment that it would be good for you to look at if you have time:

- [x] Typesystem - as I am not familiar with the type system I've not too sure how to do the correct types for things like `units` and `pmin` etc. this is causing some errors in the tests. e.g. allowing complex numbers
- [x] html representation - this is causing some tests to fail now that I have changed the `repr` to include `pmin` and `pmax` rather than `p1` and `p2`. I don't know all the html code off of the top of my head so I'm hoping it will be a quicker fix for you but let me know if not.

### CODE REVIEW

File: discretisedfield/tests/test_region.py
Comment (lang-m) [inline discretisedfield/tests/test_region.py:76]: use np.allclose

File: discretisedfield/tests/test_region.py
Comment (lang-m) [inline discretisedfield/tests/test_region.py:88]: np.allclose

File: discretisedfield/tests/test_region.py
Comment (lang-m) [inline discretisedfield/tests/test_region.py:303]: I would suggest to test some non-default values here (and also the default if not provided explicitly.

File: discretisedfield/tests/test_region.py
Comment (lang-m) [inline discretisedfield/tests/test_region.py:328]: Test non-default values

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py:502]: @lang-m how do you think is best to go about this because `type(region.pmin)=np.ndarray` and we currently only pass region to jinja2 so would we have to have a property of region which is a tuple of pmin?

File: discretisedfield/plotting/mpl_region.py
Comment (samjrholt) [reply discretisedfield/plotting/mpl_region.py]: @lang-m can you check to see if you are happy with the suggested changes

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py]: @lang-m can you check to the new generation of defaults

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py]: @lang-m I have created a new property for this

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py:126]: @lang-m I have created new properties for this and pmin and pmax

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py]: @lang-m I have added tolerances to the `allclose` method

**Comment by samjrholt** [comment]
@lang-m I think I have resolved all of the major parts. Take a look when you get a chance and let me know. The main bit I'm still unsure how to do is the best way to turn pmin into a tuple for jinja2.

File: discretisedfield/plotting/mpl_region.py
Comment (lang-m) [inline discretisedfield/plotting/mpl_region.py:122]: We have (and had before) two spaces between the axis name and the unit. Any idea if this is intentional?

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:98]:
```suggestion
        if len(p1) != len(p2):
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:101]:
```suggestion
                f" {len(p1)=} and {len(p2)=}."
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:118]: Okay for now, but we have to think about defaults and conventions for n dimensions.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:121]:
```suggestion
            units = ["m"] * ndim
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:91]:
```suggestion
            raise TypeError("p1 can only contain elements of type numbers.Number.")
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:95]:
```suggestion
            raise TypeError("p2 can only contain elements of type numbers.Number.")
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:190]: Do we really want to allow setting pmin (and pmax)? I would remove it and only allow changes via scaling and translation (which both create new objects).

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:248]: Consider removing

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:250]: I think this can't change after we create the region. So we could use `functools.cached_property`.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:313]:
```suggestion
            raise TypeError("dims can only contain elements of type str.")
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:309]: Allow None here

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:118]: Move this to `@dims.setter`

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:356]:
```suggestion
            raise TypeError("units can only contain elements of type str.")
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:352]: Allow None to reset to the default

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:121]: Move to `units.setter`

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:361]:
```suggestion
                f" Not {len(units)=} and {self.ndim=}."
```

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:466]: I will think about __repr__ and _repr_html_ and fix it when everything else is done. Please just ignore it for now.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:841]: This looks good, we have to extend the documentation.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:964]:
```suggestion
        """Convert region object to dict.
```

File: discretisedfield/tests/test_region.py
Comment (lang-m) [inline discretisedfield/tests/test_region.py:382]: I would not allow this.

**Review by lang-m — CHANGES_REQUESTED** [review]

**Comment by lang-m** [comment]
@samjrholt Please ignore the (html) representation. I will fix this once we are done with the rest.

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py:837]: @lang-m what are you thoughts on this being `isclose` rather than `allclose`?

File: discretisedfield/plotting/mpl_region.py
Comment (samjrholt) [reply discretisedfield/plotting/mpl_region.py:122]: I have no idea if this is intentional or not.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:383]:
```suggestion
    @property
```

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:316]: `edges` could change (with scaling) -> no caching
[similar for e.g. center, ... below]

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:772]:
```suggestion
            ``True`` if two regions are equal (within floating-point accuracy) and ``False`` otherwise.
```

File: discretisedfield/plotting/mpl_region.py
Comment (lang-m) [reply discretisedfield/plotting/mpl_region.py:122]: Then I would say we remove it.
```suggestion
        unit_x = rf"({uu.rsi_prefixes[multiplier]}{self.region.units[0]})"
        unit_y = rf"({uu.rsi_prefixes[multiplier]}{self.region.units[1]})"
        unit_z = rf"({uu.rsi_prefixes[multiplier]}{self.region.units[2]})"
        ax.set(xlabel=f"x {unit_x}", ylabel=f"y {unit_y}", zlabel=f"z {unit_z}")
```

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:803]: Do you mean adding an `isclose` method? Or renaming this method?

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py:316]: Scaling through multiplication and translation will return a new object. Surly as we cannot change pmin and pmax the edges and centre can never change.

File: discretisedfield/region.py
Comment (samjrholt) [reply discretisedfield/region.py:803]: Renaming this method. We are checking if one region is close to another region (similar to `np.isclose`). Whereas we would use `allclose` in field as it checks all of the values (similar to `np.allclose`)

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:316]: That is true unless we implement things like `+=` in a way that it directly changes pmin and pmax internally (instead of re-assigning the Region object in the end). Thoughts?

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:803]: But we check pmin and pmax of the region and return a single bool. For `isclose` we could also expect individual boolean values for pmin and pmax. I think I would stick to `allclose`.

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:316]: And I am still considering adding an optional argument `inplace=True` (defaulting to `False`) to avoid creating a new object if desired.

Comment (samjrholt) [reply discretisedfield/region.py:316]: Yeah that is fair enough, they are simple operations so will be a negligible time to compute anyway.

File: discretisedfield/plotting/mpl_region.py
Comment (lang-m) [inline discretisedfield/plotting/mpl_region.py:89]:
```suggestion
            raise ValueError("Only 3-dimensional regions can be plotted.")
```

File: discretisedfield/plotting/k3d_region.py
Comment (lang-m) [inline discretisedfield/plotting/k3d_region.py:57]:
```suggestion
            raise ValueError("Only 3-dimensional regions can be plotted.")
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:96]:
```suggestion
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:110]:
```suggestion
        if len(p1) != 3:
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:204]:
```suggestion
        1. Getting number of dimensions of the region.
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:200]:
```suggestion
            Number of dimensions of the region.
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:192]:
```suggestion
        r"""Number of dimensions.
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:215]:
```suggestion
        .. seealso:: :py:func:`~discretisedfield.Region.dims`
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:242]:
```suggestion

.. seealso:: :py:func:`~discretisedfield.Region.ndim`
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:226]:
```suggestion
        tuple of str
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:268]:
```suggestion
        self._dims = tuple(dims)
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:276]:
```suggestion
        tuple of str
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:314]:
```suggestion
        self._units = tuple(units)
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:350]: Should we keep the `centre` property as an alias?

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:451]:
```suggestion
        and maximum coordinate points, the same units, and the same dimension names.
```

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:117]: We should probably make `tolerance_factor` a property.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:107]: This should go to the property setter.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:361]: @samjrholt Would the following work (I think it should):
```suggestion
        return self.pmax - self.pmin
```

**Review by lang-m — APPROVED** [review]
All the changes look good.

The two open points are:
- [x] keeping `centre` as an alias?
- [x] simplifying some of the maths with pmin/pmax (see comment above)?

@swapneelap It would be good to get a second review from you. I looked at this too many times and are quite biased by now. I think it is good to be merged. But a fresh pair of eyes would be good.

**Comment by lang-m** [comment]
@samjrholt Some of the doctests are outdated and have to be fixed. And we have to update the notebook (which we may consider doing later in a separate PR, thoughts @fangohr ).

**Comment by lang-m** [comment]
Please do not change anything related to `__mul__` and `__truediv__` in this PR.

File: discretisedfield/region.py
Comment (swapneelap) [reply discretisedfield/region.py:364]: It would be good to keep the alias with a deprecation warning as we want to ultimately move towards American English.

File: discretisedfield/region.py
Comment (swapneelap) [inline discretisedfield/region.py:824]: Should we not raise an error if `other` is not a `Region` object?

**Review by swapneelap — APPROVED** [review]
Everything looks good to me. I am in favour of both the open points. I have left a few comments/questions, however, I will be happy if the PR is merged as is.

**Comment by swapneelap** [comment]
@samjrholt and @lang-m Should we also need a dictionary that maps/relates `dims` to `indices` of the, for example, `pmin` numpy array?

**Comment by lang-m** [comment]
I also briefly considered adding such a dictionary. But I think there is not too much benefit from it and it would mostly make the code more complicated. I would rely on the order as we now (implicitly) do.

File: discretisedfield/region.py
Comment (lang-m) [inline discretisedfield/region.py:402]: `FutureWarning` or `DeprecationWarning`?

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:833]: I agree, for `allclose` a `TypeError` would be better than returning False. I'll change it. Do you agree @samjrholt ?

**Comment by lang-m** [comment]
Before merging we have to go through all the failing tests and fix everything related to the changes in Region.

File: discretisedfield/region.py
Comment (swapneelap) [reply discretisedfield/region.py:402]: I feel `DeprecationWarning` is stronger than `FutureWarning`. If `centre` will be removed "soon", then I would go for `DeprecationWarning`. But I am not sure if everyone looks at it that way?

**Comment by samjrholt** [comment]
@lang-m I have fixed most of the tests involving region. Please could you look at the errors involving saving/loading

File: discretisedfield/region.py
Comment (lang-m) [reply discretisedfield/region.py:402]: We had an in-person discussion and decided to raise an exception right away.

File: discretisedfield/field.py
Comment (lang-m) [inline discretisedfield/field.py:3356]:
```suggestion
        ``self.mesh.region.pmax`` are stored as ``cell``, ``pmin``, and ``pmax``
```

File: discretisedfield/io/hdf5.py
Comment (lang-m) [inline discretisedfield/io/hdf5.py:65]:
```suggestion
        gregion.create_dataset("pmin", data=field.mesh.region.pmin)
        gregion.create_dataset("pmax", data=field.mesh.region.pmax)
```

**Pull request merged**
